{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "PPRCODE.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyN20MIF4rEHk2ydnbDLkg5n",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/YaoYinYing/PPRCODE_Guideline/blob/master/PPRCODE.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R-da9fTk4awf"
      },
      "source": [
        "# PPRCODE on Colab \n",
        "This is a experimentally reimplement work created by Yinying Yao.\n",
        "\n",
        "Original webserver (MAY not work): http://yinlab.hzau.edu.cn/pprcode/\n",
        "\n",
        "**If you find this tool useful to your research, please cite the following**:\n",
        "\n",
        "- Junjie Yan, Yinying Yao, Sixing Hong, Yan Yang, Cuicui Shen, Qunxia Zhang, Delin Zhang, Tingting Zou, Ping Yin; Delineation of pentatricopeptide repeat codes for target RNA prediction, Nucleic Acids Research. PMID:30753696 DOI: 10.1093/nar/gkz075\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UwAOeWVw4XDb",
        "cellView": "form"
      },
      "source": [
        "#@title Install packages\n",
        "import os\n",
        "import glob\n",
        "import pathlib\n",
        "import sys\n",
        "import time\n",
        "import traceback\n",
        "import pickle\n",
        "import re\n",
        "from IPython.utils import io\n",
        "import subprocess\n",
        "import tqdm.notebook\n",
        "\n",
        "try:\n",
        "  from google.colab import files\n",
        "  IN_COLAB = True\n",
        "except:\n",
        "  IN_COLAB = False\n",
        "\n",
        "\n",
        "PROSITE_URL='https://ftp.expasy.org/databases/prosite/'\n",
        "PS_SCAN_URL='https://ftp.expasy.org/databases/prosite/ps_scan/ps_scan_linux_x86_elf.tar.gz'\n",
        "PS_MOTIF='PS51375'\n",
        "TQDM_BAR_FORMAT = '{l_bar}{bar}| {n_fmt}/{total_fmt} [elapsed: {elapsed} remaining: {remaining}]'\n",
        "\n",
        "TMP_DIR = \"tmp\"\n",
        "JOBS_DIR='jobs'\n",
        "RES_DIR='results'\n",
        "RES_DIR_SCAN=f\"{RES_DIR}/scan\"\n",
        "RES_DIR_PICKLE=f'{RES_DIR}/pickle'\n",
        "RES_DIR_REPORT=f'{RES_DIR}/report'\n",
        "\n",
        "\n",
        "\n",
        "# if not already installed\n",
        "total = 10\n",
        "with tqdm.notebook.tqdm(total=total, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "  for dir in [TMP_DIR,JOBS_DIR,RES_DIR,RES_DIR_SCAN,RES_DIR_PICKLE,RES_DIR_REPORT]:\n",
        "    os.makedirs(dir, exist_ok=True)\n",
        "  pbar.update(3)\n",
        "  os.system(f\"wget -qnc {PS_SCAN_URL} -P {TMP_DIR}\")\n",
        "  pbar.update(1)\n",
        "  os.system(f'mkdir {TMP_DIR}/bin && cd {TMP_DIR}/bin && tar zxf ../ps_scan_linux_x86_elf.tar.gz')\n",
        "  pbar.update(1)\n",
        "  os.system(f'mkdir {TMP_DIR}/profiles')\n",
        "  pbar.update(1)\n",
        "\n",
        "  for p in ['profile.txt','prosite.dat','prorule.dat','evaluator.dat']:\n",
        "    os.system(f\"wget -qnc {PROSITE_URL}{p} -P {TMP_DIR}/profiles/\")\n",
        "  pbar.update(1)\n",
        "\n",
        "  try:\n",
        "    from Bio import SeqIO, Seq\n",
        "  except:\n",
        "    os.system(f\"pip3 install biopython\")\n",
        "    from Bio import SeqIO,Seq\n",
        "  pbar.update(1)\n",
        "\n",
        "  try:\n",
        "    import xlsxwriter as xw\n",
        "  except:\n",
        "    os.system(f\"pip3 install xlsxwriter\")\n",
        "  import xlsxwriter as xw\n",
        "  pbar.update(1)\n",
        "\n",
        "  if f\"{pathlib.Path(TMP_DIR).resolve()}/bin/ps_scan/\" not in os.environ['PATH']:\n",
        "    os.environ['PATH'] += f\":{TMP_DIR}/bin/ps_scan\"\n",
        "  pbar.update(1)\n",
        "\n",
        "# pprcode to rna base reflection\n",
        "codeset = [\"ND\", \"NN\", \"TD\", \"SD\", \"TN\", \"SN\", \"NS\", \"NT\", \"GN\", \"TS\", \"GD\", \"TT\", \"RD\", \"SS\", \"AD\", \"KD\", \"ST\", \"NE\",\n",
        "           \"DS\", \"CD\", \"DD\", \"NG\", \"AN\", \"GS\", \"ES\", \"LD\", \"SG\", \"GT\", \"HD\", \"CN\", \"VD\", \"ID\", \"SE\", \"NC\", \"AS\", \"RS\",\n",
        "           \"MD\", \"NK\", \"NL\", \"DT\", \"RN\", \"TG\", \"GG\", \"LN\", \"LS\", \"VN\", \"TE\", \"SP\", \"KN\", \"NR\", \"DN\", \"CS\", \"QD\", \"NV\",\n",
        "           \"ED\", \"IN\", \"VS\", \"HN\", \"AT\", \"NH\", \"AG\", \"NI\", ]\n",
        "\n",
        "# updated at 2018/12/20 for pulication, experimentally validated.\n",
        "code2rna_multi = {\"ND\": \"U>C>G\", \"NN\": \"C>U\", \"TD\": \"G>A>U\", \"SD\": \"G>>C\", \"TN\": \"A>G\", \"SN\": \"A\", \"NS\": \"C>U>A\",\n",
        "                  \"NT\": \"C>U\", \"GN\": \"A>C\", \"TS\": \"A>G\", \"GD\": \"G\", \"TT\": \"A>G\", \"RD\": \"-\", \"SS\": \"A\", \"AD\": \"G\",\n",
        "                  \"KD\": \"-\", \"ST\": \"A\", \"NE\": \"U>C\", \"DS\": \"-\", \"CD\": \"G\", \"DD\": \"-\", \"NG\": \"C>U>A\", \"AN\": \"A>>U\",\n",
        "                  \"GS\": \"-\", \"ES\": \"-\", \"LD\": \"-\", \"SG\": \"A>G\", \"GT\": \"C\", \"HD\": \"-\", \"CN\": \"A>>G\", \"VD\": \"G\",\n",
        "                  \"ID\": \"-\", \"SE\": \"-\", \"NC\": \"U>C>>A\", \"AS\": \"-\", \"RS\": \"-\", \"MD\": \"-\", \"NK\": \"C>U\", \"NL\": \"C\",\n",
        "                  \"DT\": \"-\", \"RN\": \"-\", \"TG\": \"A>>G\", \"GG\": \"-\", \"LN\": \"-\", \"LS\": \"-\", \"VN\": \"A\", \"TE\": \"G\", \"SP\": \"-\",\n",
        "                  \"KN\": \"C>U\", \"NR\": \"-\", \"DN\": \"C\", \"CS\": \"A\", \"QD\": \"-\", \"NV\": \"C>U\", \"ED\": \"-\", \"IN\": \"-\",\n",
        "                  \"VS\": \"-\", \"HN\": \"-\", \"AT\": \"-\", \"NH\": \"U>C\", \"AG\": \"-\", \"NI\": \"C\", }\n",
        "\n",
        "def ps_scan(f,s):\n",
        "  \n",
        "    if debug: print(s.id)\n",
        "    sequence_=s\n",
        "    #print(sequence_)\n",
        "    ps_out_file=f\"{RES_DIR_SCAN}/{s.id}_PPR_PSSCAN_pff.log\"\n",
        "    cmd=f\"perl {pathlib.Path(TMP_DIR).resolve()}/bin/ps_scan/ps_scan.pl -e {s.id} -p {PS_MOTIF} -o pff -d {pathlib.Path(TMP_DIR).resolve()}/profiles/prosite.dat {f} | sort -n -k 2 > {ps_out_file}\"\n",
        "    os.system(cmd)\n",
        "    ps_out_file_scan=f\"{RES_DIR_SCAN}/{s.id}_PPR_PSSCAN_scan.log\"\n",
        "    cmd=f\"perl {pathlib.Path(TMP_DIR).resolve()}/bin/ps_scan/ps_scan.pl -e {s.id} -p {PS_MOTIF} -o scan -d {pathlib.Path(TMP_DIR).resolve()}/profiles/prosite.dat {f} | sort -n -k 2 > {ps_out_file_scan}\"\n",
        "    os.system(cmd)\n",
        "    try:\n",
        "        \n",
        "        result_response_pff=open(ps_out_file,'r').read()\n",
        "        matchpos = re.findall(r'(\\d+)\\t(\\d+)\\t%s' % PS_MOTIF, result_response_pff)\n",
        "        #print(matchpos)\n",
        "        refinedpos = [[], []]\n",
        "\n",
        "        # print(matchpos)\n",
        "        match_motif = []\n",
        "\n",
        "        # get original match motifs\n",
        "        for pos in matchpos:\n",
        "            mypos=sequence_[int(pos[0])-1:int(pos[1])]\n",
        "            mypos.id=f'{s.id}_{int(pos[0])-1}_{pos[1]}'\n",
        "            #print(mypos)\n",
        "            match_motif.append(mypos)\n",
        "\n",
        "        #print(match_motif)\n",
        "        # find the motif sequence deletion by ProSite and refine the positions\n",
        "        \n",
        "        result_response_scan=open(ps_out_file_scan,'r').read()\n",
        "        for origin_motif, origin_pos in zip([str(x.seq) for x in match_motif], matchpos):\n",
        "            try:\n",
        "                if debug: print(origin_motif)\n",
        "                if fix_gap: \n",
        "                  motif_delete_find = re.findall(r'' + origin_motif + '(-+)', result_response_scan)[0]\n",
        "                else:\n",
        "                  motif_delete_find=''\n",
        "                if debug: print(\"orginal Motif= %s\\t motif deletion=%s\\t deletion length=%s\" % (origin_motif, motif_delete_find,len(motif_delete_find)))\n",
        "                refinedpos[0].append(int(origin_pos[0]))\n",
        "\n",
        "                # V1.3.1 solve the motif boundary overlap\n",
        "                if (len(refinedpos[1]) > 0 and refinedpos[1][-1] > refinedpos[0][-1]):\n",
        "                    if debug: print(\"==>refine pos 0 = %s\\t refine pos 1 = %s\" % (refinedpos[0][-1], refinedpos[1][-1]))\n",
        "                    refinedpos[1][-1] = refinedpos[0][-1]\n",
        "                    if debug: print(\"-->refine pos 0 = %s\\t refine pos 1 = %s\" % (refinedpos[0][-1], refinedpos[1][-1]))\n",
        "\n",
        "\n",
        "                refinedpos[1].append(int(origin_pos[1]) + 1 + len(motif_delete_find))\n",
        "\n",
        "            except Exception as e:\n",
        "                #traceback.print_exc()\n",
        "                #if debug: print(\"orginal Motif= %s\\t motif deletion=%s\" % (origin_motif, \"deletion not found.\"))\n",
        "                refinedpos[0].append(int(origin_pos[0]))\n",
        "                refinedpos[1].append(int(origin_pos[1]) + 1)\n",
        "\n",
        "\n",
        "\n",
        "        # the second round of refinement of position: the conserved \"P\" and the 36 aa length motif  # the conserved \"P\" label is based on weblogo results\n",
        "        # Eureka!\n",
        "        refindpos_end_2nd = []\n",
        "        try:\n",
        "            pos_index = 0\n",
        "            for refinedpos_start_slide, refinedpos_end_slide in zip(refinedpos[0], refinedpos[1]):\n",
        "                if debug: print(\"refined pos[%s]: %s -> %s \" % (\n",
        "                    pos_index, refinedpos_start_slide, refinedpos_end_slide))\n",
        "                if debug: print(sequence[refinedpos_start_slide:refinedpos_end_slide])\n",
        "                if debug: print(\"Scanning the %s/%s motif...\" % (pos_index + 1, refinedpos[0].__len__()))\n",
        "                if pos_index < (refinedpos[0].__len__() - 1):\n",
        "                    try:\n",
        "                      if (refinedpos[0][pos_index + 1] - refinedpos[1][\n",
        "                          pos_index] == 1 and\n",
        "                              sequence[refinedpos_start_slide:refinedpos_end_slide][-1] == \"P\"):\n",
        "                          if debug: print(\"Find P!\")\n",
        "                          if debug: print(len(sequence[refinedpos_start_slide:refinedpos_end_slide]))\n",
        "                          if debug: print(sequence[refinedpos_start_slide:refinedpos_end_slide][-1])\n",
        "                          refindpos_end_2nd.append(refinedpos_end_slide + 1)\n",
        "                          pass\n",
        "                      else:\n",
        "                          # print(\"No P find!\")\n",
        "                          refindpos_end_2nd.append(refinedpos_end_slide)\n",
        "                    except:\n",
        "                        refindpos_end_2nd.append(refinedpos_end_slide)\n",
        "                elif pos_index == (refinedpos[0].__len__() - 1):\n",
        "                    if sequence[refinedpos_start_slide:refinedpos_end_slide][-1] == \"P\":\n",
        "                        if debug: print(\"Find P in the last motif...\")\n",
        "                        if debug: print(len(sequence[refinedpos_start_slide:refinedpos_end_slide]))\n",
        "                        if debug: print(sequence[refinedpos_start_slide:refinedpos_end_slide][-1])\n",
        "                        refindpos_end_2nd.append(refinedpos_end_slide + 1)\n",
        "\n",
        "                    else:\n",
        "                        if debug: print(\"No P find in the last motif...\")\n",
        "                        refindpos_end_2nd.append(refinedpos_end_slide)\n",
        "                pos_index += 1\n",
        "\n",
        "\n",
        "\n",
        "        except Exception as e:\n",
        "            print(\"Oops! Something wrong hanpens with the following sequence!\")\n",
        "            print(f\"--->{s.id}\")\n",
        "            traceback.print_exc()\n",
        "        finally:\n",
        "            if debug: print(refindpos_end_2nd)\n",
        "\n",
        "        # print(match_motif)\n",
        "        # print(refinedpos)\n",
        "\n",
        "        # extract motifs scores according to the match positions\n",
        "        motif_score = []\n",
        "        \n",
        "        for line in open(ps_out_file,'r'):\n",
        "          motif_score.append(float(line.split(\"\\t\")[-2]))\n",
        "        # print(motif_score)\n",
        "        parsed_ppr=[]\n",
        "        for pos_start,pos_end,score in zip(refinedpos[0], refindpos_end_2nd,motif_score):\n",
        "            mypos=sequence_[pos_start:pos_end]\n",
        "            mypos.id=f'{s.id}_{pos_start+1}_{pos_end}'\n",
        "            mypos.annotations={\n",
        "                \"PS_SCORE\"      :   score,\n",
        "                \"PPR_STARTS\"    :   pos_start+1,\n",
        "                \"PPR_ENDS\"      :   pos_end,\n",
        "                \"FIFTH_AA\"      :   sequence_[pos_start+4],\n",
        "                \"LAST_AA\"       :   sequence_[pos_end-1],\n",
        "                \"PPR_CODE\"      :   sequence_[pos_start+4]+sequence_[pos_end-1],\n",
        "                \"RNA_BASE\"      :   code2rna_multi[sequence_[pos_start+4]+sequence_[pos_end-1]] if sequence_[pos_start+4]+sequence_[pos_end-1] in code2rna_multi.keys() else '?',\n",
        "                \n",
        "                               }\n",
        "            parsed_ppr.append(mypos)\n",
        "        return parsed_ppr\n",
        "    except Exception as e:\n",
        "        print(\"Oops! Something wrong hanpens with the following sequence!\")\n",
        "        print(f\"--->{s.id}\")\n",
        "        traceback.print_exc()\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZdTziSdN7yT_",
        "cellView": "form"
      },
      "source": [
        "#@title Enter the amino acid sequence to predict ⬇️\n",
        "import glob\n",
        "\n",
        "# define sequence\n",
        "jobname = \"ZmPPR10\" #@param {type:\"string\"}\n",
        "sequence = 'MEATGRGLFPNKPTLPAGPRKRGPLLPAAPPPPSPSSLPLDSLLLHLTAPAPAPAPAPRRSHQTPTPPHSFLSPDAQVLVLAISSHPLPTLAAFLASRRDELLRADITSLLKALELSGHWEWALALLRWAGKEGAADASALEMVVRALGREGQHDAVCALLDETPLPPGSRLDVRAYTTVLHALSRAGRYERALELFAELRRQGVAPTLVTYNVVLDVYGRMGRSWPRIVALLDEMRAAGVEPDGFTASTVIAACCRDGLVDEAVAFFEDLKARGHAPCVVTYNALLQVFGKAGNYTEALRVLGEMEQNGCQPDAVTYNELAGTYARAGFFEEAARCLDTMASKGLLPNAFTYNTVMTAYGNVGKVDEALALFDQMKKTGFVPNVNTYNLVLGMLGKKSRFTVMLEMLGEMSRSGCTPNRVTWNTMLAVCGKRGMEDYVTRVLEGMRSCGVELSRDTYNTLIAAYGRCGSRTNAFKMYNEMTSAGFTPCITTYNALLNVLSRQGDWSTAQSIVSKMRTKGFKPNEQSYSLLLQCYAKGGNVAGIAAIENEVYGSGAVFPSWVILRTLVIANFKCRRLDGMETAFQEVKARGYNPDLVIFNSMLSIYAKNGMYSKATEVFDSIKRSGLSPDLITYNSLMDMYAKCSESWEAEKILNQLKCSQTMKPDVVSYNTVINGFCKQGLVKEAQRVLSEMVADGMAPCAVTYHTLVGGYSSLEMFSEAREVIGYMVQHGLKPMELTYRRVVESYCRAKRFEEARGFLSEVSETDLDFDKKALEAYIEDAQFGR' #@param {type:\"string\"}\n",
        "sequence=sequence.upper()\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown OR upload your sequences in FASTA format\n",
        "\n",
        "#@markdown Click the icon like a folder in left panel, upload your file into a subfolder named **jobs**.\n",
        "\n",
        "#@markdown After uploading is over you must click the _Play Button_ ▶️ on each cell to continue the process.\n",
        "\n",
        "#@markdown **this will ignore any input in boxes above.**\n",
        "\n",
        "upload_my_fasta = True #@param {type:\"boolean\"}\n",
        "\n",
        "generate_excel_report=True #@param {type:\"boolean\"}\n",
        "download_full_report=True #@param {type:\"boolean\"}\n",
        "#@markdown - `upload_my_fasta` - analysis uploaded file\n",
        "#@markdown - `generate_excel_report` - generate a human-friendly report file in xlsx format\n",
        "#@markdown - `download_full_report` - once the processing is finished, the full result will be zipped and downloaded automatically.\n",
        "#@markdown ---\n",
        "debug = False #@param {type:\"boolean\"}\n",
        "#@markdown - `debug` - Some debug messages\n",
        "\n",
        "#@markdown ---\n",
        "#@markdown Experimental (manually correction to scan results)\n",
        "\n",
        "fix_gap = False #@param {type:\"boolean\"}\n",
        "#@markdown - `fix_gap` - fix gap in sequence scanning results. Turn it off so that the results will not be weird.\n",
        "\n",
        "if not upload_my_fasta:\n",
        "  open(f\"./{JOBS_DIR}/{jobname}.fasta\",'w').write(f'>{jobname}\\n{sequence}')\n",
        "else:\n",
        "  if len(glob.glob(f'{JOBS_DIR}/*.fasta'))==0:\n",
        "    raise FileNotFoundError(\"NO FASTA file is found!\")\n"
      ],
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LQ5Kdj0pCzAV",
        "cellView": "form"
      },
      "source": [
        "#@title Process the FASTAs\n",
        "fastas=glob.glob(f'{JOBS_DIR}/*.fasta')\n",
        "if debug: print(fastas)\n",
        "\n",
        "job_list=[s for f in fastas for s in SeqIO.parse(f,'fasta') ]\n",
        "if debug: print(len(job_list))\n",
        "total=len(job_list)\n",
        "\n",
        "with tqdm.notebook.tqdm(total=total, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "  for f in fastas:\n",
        "    for s in SeqIO.parse(f,'fasta'):\n",
        "      ppr=ps_scan(f,s)\n",
        "      pickle.dump(ppr,open(f'{RES_DIR_PICKLE}/{s.id}.pkl','wb'))\n",
        "      pbar.update(1)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jduRSMFAh5HO",
        "cellView": "form"
      },
      "source": [
        "#@title Generate report and download it! \n",
        "def generate_full_report():\n",
        "  \n",
        "  pkls_files=glob.glob(f'{RES_DIR_PICKLE}/*.pkl')\n",
        "  process_id = time.strftime(\"%Y%m%d_%H%M%S\", time.localtime())\n",
        "  xlsx_filename = f\"{RES_DIR_REPORT}/PPRCODE_RESULTS_{process_id}.xlsx\" \n",
        "  test_book = xw.Workbook(xlsx_filename)\n",
        "\n",
        "  # set workbook property\n",
        "  test_book.set_properties({\n",
        "                'title': 'PPRCODE PREDICTION',\n",
        "                'subject': 'http://yinlab.hzau.edu.cn/pprcode/',\n",
        "                'author': 'PPRCODE project',\n",
        "                'manager': 'Yan et al.',\n",
        "                'company': 'Yin Lab',\n",
        "                'comments': 'Created with pprcode project.'})\n",
        "\n",
        "  for pf in pkls_files:\n",
        "    data = pickle.load(open(pf,'rb'))\n",
        "    motif_starts  = [x.annotations[  \"PPR_STARTS\"   ] for x in data]\n",
        "    motif_ends    = [x.annotations[  \"PPR_ENDS\"     ] for x in data]\n",
        "    motif_sequence= [str(x.seq) for x in data]\n",
        "    fifth_code    = [x.annotations[  \"FIFTH_AA\"     ] for x in data]\n",
        "    last_code     = [x.annotations[  \"LAST_AA\"      ] for x in data]\n",
        "    ppr_code      = [x.annotations[  \"PPR_CODE\"     ] for x in data]\n",
        "    rna_base      = [x.annotations[  \"RNA_BASE\"     ] for x in data]\n",
        "    motif_length  = [len(str(x.seq)) for x in data]\n",
        "    motif_score   = [x.annotations[  \"PS_SCORE\"   ] for x in data]\n",
        "    data_list=[[a,b,c,d,e,f,g,h,i] \n",
        "               for a,b,c,d,e,f,g,h,i in zip(\n",
        "                   motif_starts,\n",
        "                   motif_ends,\n",
        "                   motif_sequence,\n",
        "                   fifth_code,\n",
        "                   last_code,\n",
        "                   ppr_code,\n",
        "                   rna_base,\n",
        "                   motif_length,\n",
        "                   motif_score)]\n",
        "    seq_id=pathlib.Path(pf).resolve().stem\n",
        "    try:\n",
        "      write_excel_tables(test_book,seq_id,data_list)\n",
        "    except:\n",
        "      print(\"Oops! Something wrong hanpens when writing in to excel report!\")\n",
        "      print(seq_id)\n",
        "  print(\"All jobs are accomplished.\")\n",
        "  test_book.close()\n",
        "\n",
        "def write_excel_tables(test_book, sequence_name, ppr_total_list_reversed):\n",
        "    # reversed ppr code result list for making xlsx file\n",
        "\n",
        "    code_count_result_name = \"PPR motifs and PPR codes: %s\" % sequence_name\n",
        "    format_title = test_book.add_format()\n",
        "    format_content = test_book.add_format()\n",
        "    format_title.set_font(\"Times New Roman\")\n",
        "    format_title.set_font_size(11)\n",
        "    format_title.set_bold()\n",
        "    format_title.set_bottom(2)\n",
        "\n",
        "    format_content.set_font(\"Times New Roman\")\n",
        "    format_content.set_font_size(11)\n",
        "\n",
        "    worksheet = test_book.add_worksheet(sequence_name)\n",
        "    worksheet.set_column(0, 1, 10)\n",
        "    worksheet.set_column(2, 2, 50)\n",
        "    worksheet.set_column(3, 8, 10)\n",
        "\n",
        "    worksheet.write(0, 0, code_count_result_name, format_title)\n",
        "    for n in range(1, 9):\n",
        "        worksheet.write(0, n, \"\", format_title)\n",
        "    worksheet.write(1, 0, \"Motif Start\", format_title)\n",
        "    worksheet.write(1, 1, \"Motif End\", format_title)\n",
        "    worksheet.write(1, 2, \"Motif Sequence\", format_title)\n",
        "    worksheet.write(1, 3, \"Fifth amino acid\", format_title)\n",
        "    worksheet.write(1, 4, \"Last amino acid\", format_title)\n",
        "    worksheet.write(1, 5, \"PPR code\", format_title)\n",
        "    worksheet.write(1, 6, \"RNA base\", format_title)\n",
        "    worksheet.write(1, 7, \"Motif Length\", format_title)\n",
        "    worksheet.write(1, 8, \"ProSite Score\", format_title)\n",
        "\n",
        "    row = 2\n",
        "    col = 0\n",
        "\n",
        "    rna = \"\"\n",
        "    for ppr_list in ppr_total_list_reversed:\n",
        "        # print(ppr_list)\n",
        "        motif_start = ppr_list[0]\n",
        "        motif_end = ppr_list[1]\n",
        "        motif_sequence = ppr_list[2]\n",
        "        fifth_code = ppr_list[3]\n",
        "        last_code = ppr_list[4]\n",
        "        ppr_code = ppr_list[5]\n",
        "        rna_base = ppr_list[6]\n",
        "\n",
        "        rna = ''.join([rna, \" (\", rna_base, \")\"])\n",
        "\n",
        "        motif_length = ppr_list[7]\n",
        "        motif_score = ppr_list[8]\n",
        "        worksheet.write(row, col, motif_start, format_content)\n",
        "        worksheet.write(row, col + 1, motif_end, format_content)\n",
        "        worksheet.write(row, col + 2, motif_sequence, format_content)\n",
        "        worksheet.write(row, col + 3, fifth_code, format_content)\n",
        "        worksheet.write(row, col + 4, last_code, format_content)\n",
        "        worksheet.write(row, col + 5, ppr_code, format_content)\n",
        "        worksheet.write(row, col + 6, rna_base, format_content)\n",
        "        worksheet.write(row, col + 7, motif_length, format_content)\n",
        "        worksheet.write(row, col + 8, motif_score, format_content)\n",
        "\n",
        "        row += 1\n",
        "        pass\n",
        "    worksheet.write(row + 1, col, \"The RNA predicted by this PPR sequence:\", format_title)\n",
        "    worksheet.write(row + 2, col, rna, format_title)\n",
        "    worksheet.write(row + 4, col, \"Notes:\", format_title)\n",
        "    worksheet.write(row + 5, col,\n",
        "                    \"'-' represents no correlated RNA bases for this PPR code identified by biochemical assays (EMSA, ITC);\",\n",
        "                    format_content)\n",
        "    worksheet.write(row + 6, col,\n",
        "                    \"'?' represents unknown bases, as no biochemical assays have been performed to identify the correlated RNA bases for the PPR code.\",\n",
        "                    format_content)\n",
        "    worksheet.write(row + 8, col, \"If you find this tool useful to your research, please cite the following:\",\n",
        "                    format_title)\n",
        "    worksheet.write(row + 9, col,\n",
        "                    \"Junjie Yan, Yinying Yao, Sixing Hong, Yan Yang, Cuicui Shen, Qunxia Zhang, Delin Zhang, Tingting Zou, Ping Yin; Delineation of pentatricopeptide repeat codes for target RNA prediction, Nucleic Acids Research, gkz075, https://doi.org/10.1093/nar/gkz075\",\n",
        "                    format_content)\n",
        "    worksheet.write(row + 11, col, \"Copyright: Yin lab\", format_title)\n",
        "    pass\n",
        "\n",
        "\n",
        "if generate_excel_report: \n",
        "  generate_full_report()\n",
        "\n",
        "zipped=f'PPRCODE_{time.strftime(\"%Y%m%d_%H%M%S\", time.localtime())}.zip'\n",
        "os.system(f\"zip -FSr {zipped} {RES_DIR}\")\n",
        "\n",
        "if download_full_report:\n",
        "  if IN_COLAB:\n",
        "    files.download(f'{zipped}')\n",
        "  else:\n",
        "    print(\"this notebook appears to be running locally, to download click folder icon to the left, navigate to file, right click and download\")\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}