

![PPRCODE](image/5i9f_MN_97c.png)
_Cover Image is presented with [**MolecularNodes Project**](https://github.com/BradyAJohnston/MolecularNodes)_

---
[![Paper-PPRCODE](https://tinyurl.com/pprcode-badge)](https://doi.org/10.1093/nar/gkz075)
[![Available at Colab](https://img.shields.io/badge/Available%20at-%20Colab-eeae3c.svg?logoWidth=16&logo=googlecolab)](https://colab.research.google.com/github/YaoYinYing/PPRCODE_Guideline/blob/main/PPRCODE.ipynb)
[![Available at BioLib](https://img.shields.io/badge/Available%20at-BioLib-55b5a6.svg?logoWidth=16&logo=data:image/x-icon;base64,iVBORw0KGgoAAAANSUhEUgAAADYAAABGCAYAAACHZ2ccAAAACXBIWXMAAAu5AAALuQHs9frzAAAExUlEQVRogeVbPXbaQBAe56U3KaQ23MDcwHQqbZeqRE4Q0qg1tDQmJ4ioKA0lleEEgRvgFjXhBM5bPKsM+6fVHyzO9x7v2VqB9tv5dmZ2d3T19vYGHxGfT83JW0x7ADAGgGsAeAWAXhqES+nGijiZxbzFtAUACQDcSY0AwzQIB9LVCjgJMW8x7QDADAC+So3/sELrbaWWEvjUKKN3UswSv3NIMdwCwNpbTO+llhJozGIovRl2uCh+AsAgDcI/ThHDUU/QQZTFBqW5LvP92qXoLabM4z1XJMVwAwBLbzHtSy0WqM1i6CAS7FDdmKP1rKVZi8UwNi0bIgUYIphj6UotGlSyWE5sagpWMa80Me999BILN94EcmNeKSlibHo5EymwiXmFLOYtpm20UpnY1BSUMc+aWE2xqSlIMS+XGDoIJr3vUqN7+JEGIYujVnNsfCGkGJ4w9FgR005QR3FvS+zS0LIlNpOuuI3Zf0+s9v2IBrHh2UguMQx8c6nBTSS8V7bO41LkmPXzIxHb0KTYitiFyDGh/2iJ+VHc9qO4Ry65brWj/imJ+VHMd5guhdhGXJspiWF+yJb5t0jSdTlKIUki5kcx2xWKyCWaK7pqtUS8cETMj2K23H8S7nGd2Ktq7zEj5kdxR9PxO0GOG+mO80LV5yOL9Q2rY2o1yexnhrI/lFhbalUTU47QmaCUoUhMeQOCynHrkBy1g0yJDfCEUQcX5ajtR0ZsNxn9wXmmg2ty1MoQRHe/m4xmhiCcbWM7Ikfj4EoBGtOovXT1PSS4JEcp26CQiKEke9Kd73BFjvs0CI3Pl4iBWZIZsTPL0UgKdMQQfYUkrwU5GuXQIMoT201GWwwBIsR5ppyPDSJXhiDu3fOT/jQIs5NDP4qXwunKfjcZtch3OjgALWgebLD7Nke2IjHmNH7RU0O2ksashOaRDzgPnYUoRS69R7SETpLO7+dnFsPzr2fSxpbbHf6PQpLDMzmPLQ62EbT6TUynbtiRLDnI7mPpEMcjfk6NoUJBEg5SRNmpjl8f8XiWSXKNP3puaPNDCj7HTHMmS512k9HgzDni0NZpcWKmuXIrlP3o0q0mwJZRE+aFAeALDqwVqPMwHUazINzhe3d+FA8aml97HGT2mdk4CR2o8xgaOnuNkuSBe4yWq6POY0WIGOcPJhCHaZMGoXF1oQvQOnzjP4hbdS+a+0zYEKsscTWhBYahLn54rRZLq4yZjlQOIRQji2BSafOUxo9im4qCPSatnIhRXuihu2gZlac+IA3CK+kigVTFzSziLaZr7IhIjkuSe9EB/i1Kck6I5MmrTYh0Fc9UwbQ3c4BkMfLAjoYcwwPNsPFUpo1ETB6Wz5MuIVN0nkpVOCpoiZFOqOoQjySZB6yU40TK1jSyZ/bznAaHkVgOuXkahMrATuZJt4C8TGAee1ykwtSqSAzJjYVTGEBZJOgcKJG6yv1K1+IXLetLFOSaQOVXQwpXmGIRpi6QV8Ueaw/HVX+oVOmsRSAvA2VBZVlUqQk2BfIiWOE+htF9F0XVKm5TrMvDKxJqZO+k8gsFJcjt0XXX+lqViFrelDDEOhETnEe1vEplQp2vgJjIrZDQyTZ/an0biRRGd8jR78A2DaoTH/PlUwD4C+B8JnMOxg8tAAAAAElFTkSuQmCC)](https://biolib.com/YaoYinYing/pprcode/)
[![Platform: docker](https://img.shields.io/badge/Platform-docker-blue.svg?logoWidth=16&logo=docker)](https://hub.docker.com/r/yaoyinying/pprcode)
[![Build with Ubuntu x86_64](https://img.shields.io/badge/Build%20in-Ubuntu%20x86__64-d85e33.svg?logoWidth=16&logo=ubuntu)](https://hub.docker.com/r/yaoyinying/pprcode)
----
[![PPRCODE Workflow](https://github.com/YaoYinYing/PPRCODE_Guideline/actions/workflows/pprcode-guideline.yml/badge.svg)](https://github.com/YaoYinYing/PPRCODE_Guideline/actions/workflows/pprcode-guideline.yml)
[![Docker Image Pulls](https://img.shields.io/docker/pulls/yaoyinying/pprcode)](https://hub.docker.com/r/yaoyinying/pprcode)
----

# PPRCODE

Original Project site: [PPR Code Prediction Server - From PPR to RNA](http://yinlab.hzau.edu.cn/pprcode/)

## NOTE
This original website is Down.

Please switch to:
 1. Colab release
 2. Docker release
 3. Biolib release
 

## Three ways to run PPRCODE

 1. [WebServer from BioLib](https://biolib.com/YaoYinYing/pprcode/); 
 **the [original webserver](http://yinlab.hzau.edu.cn/pprcode/) provided by [Yin Lab](http://yinlab.hzau.edu.cn/) is down and will be no longer maintained.**
 2. [Colab Reimplementation](https://colab.research.google.com/github/YaoYinYing/PPRCODE_Guideline/blob/main/PPRCODE.ipynb)
 3. Local run: Docker image or BioLib cloud scripts
 
## Run PPRCODE locally via APIs provided by BioLib
 1. install required BioLib package
    ```shell
    pip3 install -U pybiolib
    ```
 2. run PPRCODE via Shell commands
    ```shell
    wget -qnc https://raw.githubusercontent.com/YaoYinYing/PPRCODE_Guideline/main/ppr_example.fasta 
    biolib run YaoYinYing/pprcode --fasta ppr_example.fasta
    ```
  **the run results will be located at $PWD/biolib_results**
  
**PS**: 
  _Due to the I/O issue of Biolib as docker container wrapper, the customized `--save_dir` option will produce no results._
 
## Run PPRCODE locally in docker
 1. Install docker daemon by following the [official getting-started page](https://www.docker.com/get-started/) instruction.
 2. Clone this repo
    ```shell
    git clone https://github.com/YaoYinYing/PPRCODE_Guideline
    ```
 3. PPRCODE docker image.

    **fetch the latest image**
    ```shell
    docker pull yaoyinying/pprcode:latest
    ```
    **You may also build it from scratch:**
    ```shell
    cd PPRCODE_Guideline
    docker build -f docker/Dockerfile -t pprcode . 
    ```
    Alternatively, if you wish to run PPRCODE with your own version of `docker/run_pprcode.py`, you may build a patched image for local usage by the following: 
    ```shell
    docker build -f docker/Dockerfile_patch -t pprcode .
    ```
 4. Create Conda environment for run this docker image in an instance container
    ```shell
    conda create -y -n pprcode python pip
    conda activate pprcode
    cd <repo>/PPRCODE_Guideline
    pip install -r docker/requirements.txt
    ```
 5. Run `run_docker.py` to an example data
    ```shell
    conda activate pprcode
    mkdir test
    
    # fetch an example dataset 
    wget -qnc https://raw.githubusercontent.com/YaoYinYing/PPRCODE_Guideline/main/ppr_example.fasta -P test
    
    # use PS_Scan as default program
    python /repo/PPRCODE_Guideline/docker/run_docker.py --fasta test/ppr_example.fasta --save_dir ./save-ps_scan  --plot_item=bar,score,edge,ppr,rna
    
    # or use pprfinder provided by Small's Lab
    python  /repo/PPRCODE_Guideline/docker/run_docker.py --fasta test/ppr_example.fasta --save_dir ./save-pprfinder --plot_item=bar,score,edge,ppr,rna --program=pprfinder
    ```
 6. Advance options
    ```shell
    python  /repo/PPRCODE_Guideline/docker/run_docker.py --help
    ```


## FAQs

### _Q_: What is PPR and PPR code?
Pentatricopeptide repeat (PPR) proteins constitute a large family whose members serve as single-stranded RNA (ssRNA)-binding proteins; these proteins are particularly abundant in terrestrial plants, as more than 400 members have been identified in Arabidopsis and rice.

PPR proteins are typically characterized by tandem degenerate repeats of a 35-amino acid motif. Within a given repeat, the combinatorial di-residues at the 5th and 35th positions are responsible for specific RNA base recognition. These di-residues are referred to as the **PPR code**. 


### _Q_: What is PPRCODE prediction server?
**PPRCODE** prediction server is aimed to provide services to the PPR community to facilitate PPR code and target RNA prediction. Once a PPR protein sequence is submitted, the server firstly identifies the PPR motifs using the PScan algorithm provided by Prosite, and then outputs the individual PPR motifs that is demarcated based on the PPR structure. PPR code is generally extracted from the 5th and 35th amino acids of each PPR motif, and the best matched RNA base for the PPR code is provided. As a result, the potential RNA target for the PPR sequence is available. 

![PPRCODE Prediction Server: Interface](image/pprcode_biolib.png)


### _Q_: How do I submit a sequence to the PPRCODE prediction server?

Go to the [PPRCODE prediction server in BioLib](https://biolib.com/YaoYinYing/pprcode/) submission form directly and do the following:
  1. Paste your FASTA sequence in the upper text area.
  2. Modify the options if needed.
  3. Click the Run button.
 After the submission, the webpage will be automatically run for several second until the job is finished. 
 
### _Q_: How long does it take to finish a task?
Less than three second for each sequence.

### _Q_: How many sequences can I submit in one submission?
As many as you want.


### _Q_: What does the prediction result mean?
The result page contains a table like the following:

> This is a demo sequence of PPR10 from *Zea mays*.


Motif Start | Motif End | Motif Sequence | Fifth amino acid | Last amino acid | PPR Code | RNA base | Motif Length | ProSite Score
 -----|-----|-----|-----|-----|----|----|----|----
138 | 172 | ASALEMVVRALGREGQHDAVCALLDETPLPPGSRL | E | L | EL | ? | 35 | 5.031
174 | 208 | VRAYTTVLHALSRAGRYERALELFAELRRQGVAPT | T | T | TT | A>G | 35 | 12.989
209 | 244 | LVTYNVVLDVYGRMGRSWPRIVALLDEMRAAGVEPD | N | D | ND | U>C>G | 36 | 11.093
245 | 279 | GFTASTVIAACCRDGLVDEAVAFFEDLKARGHAPC | S | C | SC | ? | 35 | 11.411
280 | 314 | VVTYNALLQVFGKAGNYTEALRVLGEMEQNGCQPD | N | D | ND | U>C>G | 35 | 12.737
315 | 349 | AVTYNELAGTYARAGFFEEAARCLDTMASKGLLPN | N | N | NN | C>U | 35 | 11.477
350 | 384 | AFTYNTVMTAYGNVGKVDEALALFDQMKKTGFVPN | N | N | NN | C>U | 35 | 14.096
385 | 419 | VNTYNLVLGMLGKKSRFTVMLEMLGEMSRSGCTPN | N | N | NN | C>U | 35 | 10.358
420 | 454 | RVTWNTMLAVCGKRGMEDYVTRVLEGMRSCGVELS | N | S | NS | C>U>A | 35 | 9.887
455 | 489 | RDTYNTLIAAYGRCGSRTNAFKMYNEMTSAGFTPC | N | C | NC | U>C>>A | 35 | 11.674
490 | 524 | ITTYNALLNVLSRQGDWSTAQSIVSKMRTKGFKPN | N | N | NN | C>U | 35 | 11.542
525 | 560 | EQSYSLLLQCYAKGGNVAGIAAIENEVYGSGAVFPS | S | S | SS | A | 36 | 6.467
561 | 595 | WVILRTLVIANFKCRRLDGMETAFQEVKARGYNPD | R | D | RD | - | 35 | 6.445
596 | 630 | LVIFNSMLSIYAKNGMYSKATEVFDSIKRSGLSPD | N | D | ND | U>C>G | 35 | 12.419
631 | 666 | LITYNSLMDMYAKCSESWEAEKILNQLKCSQTMKPD | N | D | ND | U>C>G | 36 | 8.67
667 | 701 | VVSYNTVINGFCKQGLVKEAQRVLSEMVADGMAPC | N | C | NC | U>C>>A | 35 | 13.778
702 | 736 | AVTYHTLVGGYSSLEMFSEAREVIGYMVQHGLKPM | H | M | HM | ? | 35 | 10.348
737 | 771 | ELTYRRVVESYCRAKRFEEARGFLSEVSETDLDFD | R | D | RD | - | 35 | 8.089


and finally you will also get a predicted sequence like this:
>(?) (A>G) (U>C>G) (?) (U>C>G) (C>U) (C>U) (C>U) (C>U>A) (U>C>>A) (C>U) (A) (-) (U>C>G) (U>C>G) (U>C>>A) (?) (-)


### _Q_: Why does the prediction result of my sequence look like a mess?
PS_Scan/PPRfinder identifies the sequence and motifs of a PPR protein by its similarity to the general P-type PPR. Sequences with low identity will hardly be predicted. In this circumstance, manual correction is strongly recommended. 

## Troubleshoot
If there is any problem and advice with the website, you are welcome to contact us via [email](mailto:yaoyy@webmail.hzau.edu.cn).

## Contributers:
* **Yinying Yao**: Main program development and further maintainance.
* **Zeyuan Guan**: Basic Framework of the original webserver.
* **Junjie Yan**: Writing and data collecting.
* **Xiang Wang**: Providing useful advices to the original webserver design.

## Cite information
Yan Junjie#, Yao Yinying#, Hong Sixing, Yang Yan, Shen Cuicui, Zhang Qunxia, Zhang Delin, Zou Tingting, Yin Ping*. Delineation of pentatricopeptide repeat codes for target RNA prediction, Nucleic Acids Research. 2019 February 11. doi: doi.org/10.1093/nar/gkz075

